### ISSUE TITLE
Use shared_ptr

### ISSUE THREAD

**Issue by mnett** [mnett/lazytensor#33]
Rewrite the whole system to handle expressions through shared pointers rather than (unsafe) references. This makes it harder to violate the expression lifetime constraints that are implicitly in place.

**Comment by Kintaro** [comment]
Shouldn't it be unique_pointer then? Use of shared_ptr is considered harmful and deprecated.

**Comment by mnett** [comment]
As a general statement that is absolute nonsense. Neither shared nor unique
pointers are deprecated (auto pointer will be). You simply choose between
shared and unique to express ownership. If several expressions need to have
guarantees on the lifetime of some expression X, a shared pointer is
exactly what we need

**Comment by Kintaro** [comment]
Ok, rephrased: use of unique_ptr is to be preferred wherever possible. Not sure if shared_ptr is really preferrable here.

**Comment by mnett** [comment]
I agree on "if you use them correctly there's not even s choice which one
to use" :)

**Comment by Kintaro** [comment]
That's why I'm unsure if shared_ptrs will help us out here. A few intermediate expressions might resolve to the same expression, but that's difficult to detect. Especially if you want to use them to trace lifetime constraints, unique_ptrs seem more like the one we might want to use, as opposed to shared_ptrs. Especially in context with MPI. shared_ptrs are more difficult to handle with MPI, regarding serialization along transfers.
